### PR TITLE
[sgd] Add support for multi-model multi-optimizer training

### PR DESCRIPTION
**Pull request by richardliaw**
Needed for DCGAN example.

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.

### CODE REVIEW

**Comment by simon-mo** [comment]
I think you will need to change the docs as well? For the renaming `dataset_creators->dataset_creator`

**Comment by richardliaw** [comment]
Ah, that change is automatically reflected in the docs (because we do a literalinclude).

File: python/ray/experimental/sgd/pytorch/examples/dcgan.py [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+        root="~/mnist/",
+        download=True,
+        transform=transforms.Compose([
+            transforms.Resize(32),
+            transforms.ToTensor(),
+            transforms.Normalize((0.5, ), (0.5, )),
+        ]))
+
+    # Create the dataloader
+    # Make this distributed
```
Comment (simon-mo) [inline python/ray/experimental/sgd/pytorch/examples/dcgan.py]: What is distributed here?

File: python/ray/experimental/sgd/pytorch/examples/dcgan.py [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+            break
+
+        netD.zero_grad()
+        real_cpu = data[0].to(device)
+        b_size = real_cpu.size(0)
+        label = torch.full((b_size, ), real_label, device=device)
+        output = netD(real_cpu).view(-1)
+        errD_real = criterion(output, label)
+        errD_real.backward()
+        # D_x = output.mean().item()
```
Comment (simon-mo) [inline python/ray/experimental/sgd/pytorch/examples/dcgan.py]: Remove?

**Review by simon-mo — APPROVED** [review]
Two minor comments. the example loos good.

**Comment by richardliaw** [comment]
jenkins test tune

**Comment by richardliaw** [comment]
jenkins test tune

**Comment by richardliaw** [comment]
jenkins test tune

**Comment by richardliaw** [comment]
jenkins test tune

**Comment by richardliaw** [comment]
jenkins test tune

**Pull request merged**